File: packages/node-client/src/audit.ts
// ADR-1006 Phase 1 Sprint 1.3 — listAuditLog + cursor iterator.

import type { HttpClient } from './http'
import type { AuditLogEnvelope, AuditLogItem } from './types'

export interface ListAuditLogInput {
  eventType?: string
  entityType?: string
  createdAfter?: string
  createdBefore?: string
  cursor?: string
  limit?: number
  traceId?: string
  signal?: AbortSignal
}

export async function listAuditLog(
  http: HttpClient,
  input: ListAuditLogInput = {},
): Promise<AuditLogEnvelope> {
  const resp = await http.request<AuditLogEnvelope>({
    method: 'GET',
    path: '/audit',
    query: {
      event_type: input.eventType,
      entity_type: input.entityType,
      created_after: input.createdAfter,
      created_before: input.createdBefore,
      cursor: input.cursor,
      limit: input.limit,
    },
    signal: input.signal,
    traceId: input.traceId,
  })
  return resp.body
}

export async function* iterateAuditLog(
  http: HttpClient,
  input: ListAuditLogInput = {},
): AsyncIterableIterator<AuditLogItem> {
  let cursor = input.cursor
  while (true) {
    const page: AuditLogEnvelope = await listAuditLog(http, { ...input, cursor })
    for (const item of page.items) yield item
    if (!page.next_cursor) return
    cursor = page.next_cursor
  }
}

File: packages/go-client/audit.go
package consentshield

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
)

// ListAuditLogParams filters the audit-log list endpoint.
type ListAuditLogParams struct {
	Action  string
	Subject string
	Cursor  string
	Limit   int
	TraceID string
}

func (p ListAuditLogParams) query() url.Values {
	q := url.Values{}
	if p.Action != "" {
		q.Set("action", p.Action)
	}
	if p.Subject != "" {
		q.Set("subject", p.Subject)
	}
	if p.Cursor != "" {
		q.Set("cursor", p.Cursor)
	}
	if p.Limit > 0 {
		q.Set("limit", fmt.Sprintf("%d", p.Limit))
	}
	return q
}

// ListAuditLog returns a single page of audit events.
func (c *Client) ListAuditLog(ctx context.Context, p ListAuditLogParams) (*AuditLogEnvelope, error) {
	resp, err := c.transport.do(ctx, &httpRequest{
		Method:  "GET",
		Path:    "/v1/audit",
		Query:   p.query(),
		TraceID: p.TraceID,
	})
	if err != nil {
		return nil, err
	}
	var out AuditLogEnvelope
	if err := json.Unmarshal(resp.Body, &out); err != nil {
		return nil, fmt.Errorf("consentshield: decode list_audit_log: %w", err)
	}
	return &out, nil
}

// AuditLogPaginator walks every page of /v1/audit.
type AuditLogPaginator struct {
	client *Client
	params ListAuditLogParams
	done   bool
	last   []AuditLogItem
	err    error
}

// IterateAuditLog returns a paginator over /v1/audit.
func (c *Client) IterateAuditLog(p ListAuditLogParams) *AuditLogPaginator {
	return &AuditLogPaginator{client: c, params: p}
}

func (it *AuditLogPaginator) Next(ctx context.Context) bool {
	if it.done {
		return false
	}
	env, err := it.client.ListAuditLog(ctx, it.params)
	if err != nil {
		it.err = err
		it.done = true
		return false
	}
	it.last = env.Items
	if env.NextCursor == nil || *env.NextCursor == "" {
		it.done = true
	} else {
		it.params.Cursor = *env.NextCursor
	}
	return true
}

func (it *AuditLogPaginator) Page() []AuditLogItem { return it.last }
func (it *AuditLogPaginator) Err() error           { return it.err }
